roachtest: tpccbench/nodes=6/cpu=16/multi-az failed

roachtest.tpccbench/nodes=6/cpu=16/multi-az [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3824348&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3824348&tab=artifacts#/tpccbench/nodes=6/cpu=16/multi-az) on master @ [a80cfbee826f70988381a0a85c0fe7aba0115484](https://github.com/cockroachdb/cockroach/commits/a80cfbee826f70988381a0a85c0fe7aba0115484):


```
The test failed on branch=master, cloud=gce:
test timed out (see artifacts for details)
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #72869 roachtest: tpccbench/nodes=6/cpu=16/multi-az failed [C-test-failure O-roachtest O-robot branch-release-21.2]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpccbench/nodes=6/cpu=16/multi-az.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Easy enough:

```
panic: child not present in parent [recovered]
        panic: child not present in parent

goroutine 367076 [running]:
panic({0x426b1c0, 0x83d8ab0})
        /usr/local/go/src/runtime/panic.go:1147 +0x3a8 fp=0xc01a5f1c08 sp=0xc01a5f1b48 pc=0x488448
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000987710, {0x84eef58, 0xc004a5f090})
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:254 +0xaa fp=0xc01a5f1c68 sp=0xc01a5f1c08 pc=0xff52aa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2·dwrap·10()
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:482 +0x2e fp=0xc01a5f1c90 sp=0xc01a5f1c68 pc=0xff676e
runtime.deferCallSave(0xc01a5f1d98, 0xc01a5f1f90)
        /usr/local/go/src/runtime/panic.go:950 +0x82 fp=0xc01a5f1ca0 sp=0xc01a5f1c90 pc=0x488042
runtime.runOpenDeferFrame(0xc001f7b9f0, 0xc000fa4000)
        /usr/local/go/src/runtime/panic.go:889 +0x27b fp=0xc01a5f1d20 sp=0xc01a5f1ca0 pc=0x4879db
panic({0x426b1c0, 0x83d8ab0})
        /usr/local/go/src/runtime/panic.go:1038 +0x215 fp=0xc01a5f1de0 sp=0xc01a5f1d20 pc=0x4882b5
github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).childFinished(0xc00d727240, 0xed9396646)
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/crdbspan.go:630 +0x38e fp=0xc01a5f1e80 sp=0xc01a5f1de0 pc=0xb941ae
github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).finish(0xc004a5ef40)
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/crdbspan.go:188 +0x145 fp=0xc01a5f1ee8 sp=0xc01a5f1e80 pc=0xb909c5
github.com/cockroachdb/cockroach/pkg/util/tracing.(*spanInner).Finish(0xc004a5ef00)
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span_inner.go:90 +0x4d fp=0xc01a5f1f18 sp=0xc01a5f1ee8 pc=0xb9b88d
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0x0)
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:83 +0x4a fp=0xc01a5f1f30 sp=0xc01a5f1f18 pc=0xb9a9ca
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2·dwrap·12()
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x26 fp=0xc01a5f1f48 sp=0xc01a5f1f30 pc=0xff66a6
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:492 +0x168 fp=0xc01a5f1fe0 sp=0xc01a5f1f48 pc=0xff65a8
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc01a5f1fe8 sp=0xc01a5f1fe0 pc=0x4bc2c1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
        /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:481 +0x445
```

@andreimatei you've touched tracing and invariants around spans recently, so I'm pretty sure you would either be the right person to fix this problem or know who is.

--- a/pkg/util/tracing/crdbspan.go
+++ b/pkg/util/tracing/crdbspan.go
@@ -587,22 +587,31 @@ func (s *crdbSpan) getRecordingNoChildrenLocked(
 	return rs
 }
 
-func (s *crdbSpan) addChild(child *crdbSpan, collectChildRec bool) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+// addChildLocked adds a child to the receiver.
+//
+// The receiver's lock must be held.
+//
+// The adding fails if the receiver is not recording  (non-recording spans don't
+// accumulate children) or if the reciver has reached the maximum number of
+// spans it can keep track off. Returns false in these cases.
+func (s *crdbSpan) addChildLocked(child *crdbSpan, collectChildRec bool) bool {
+	s.mu.AssertHeld()
 	if s.recordingType() == RecordingOff {
 		// We're not recording; there's nothing to do. The caller also checks the
 		// recording status, but we want to also check it here under the lock
 		// (double-checked locking).
-		return
+		return false
 	}
 
-	if len(s.mu.recording.openChildren) < maxChildrenPerSpan {
-		s.mu.recording.openChildren = append(
-			s.mu.recording.openChildren,
-			childRef{crdbSpan: child, collectRecording: collectChildRec},
-		)
+	if len(s.mu.recording.openChildren) >= maxChildrenPerSpan {
+		return false
 	}
+
+	s.mu.recording.openChildren = append(
+		s.mu.recording.openChildren,
+		childRef{crdbSpan: child, collectRecording: collectChildRec},
+	)
+	return true
 }
 
 // childFinished is called when a child is Finish()ed. Depending on the

--- a/pkg/util/tracing/span_test.go
+++ b/pkg/util/tracing/span_test.go
@@ -350,13 +350,18 @@ func TestSpanMaxChildren(t *testing.T) {
 	tr := NewTracer()
 	sp := tr.StartSpan("root", WithRecording(RecordingStructured))
 	defer sp.Finish()
-	for i := 0; i < maxChildrenPerSpan+123; i++ {
-		tr.StartSpan(fmt.Sprintf("child %d", i), WithParent(sp))
+	numChildren := maxChildrenPerSpan + 123
+	children := make([]*Span, numChildren)
+	for i := 0; i < numChildren; i++ {
+		children[i] = tr.StartSpan(fmt.Sprintf("child %d", i), WithParent(sp))
 		exp := i + 1
 		if exp > maxChildrenPerSpan {
 			exp = maxChildrenPerSpan
 		}
 		require.Len(t, sp.i.crdb.mu.recording.openChildren, exp)
+	}
+	for _, s := range children {
+		s.Finish()
 	}
 }
 

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -740,9 +740,16 @@ func (t *Tracer) startSpanGeneric(
 		// created Span to the parent so that the parent will later be able to
 		// collect the child's recording.
 		if opts.Parent != nil && opts.Parent.i.crdb != nil {
-			if opts.Parent.i.crdb.recordingType() != RecordingOff {
-				s.i.crdb.mu.parent = opts.Parent.i.crdb
-				opts.Parent.i.crdb.addChild(s.i.crdb, !opts.ParentDoesNotCollectRecording)
+			parent := opts.Parent.i.crdb
+			if parent.recordingType() != RecordingOff {
+				// We're going to hold the parent's lock while we link both the parent
+				// to the child and the child to the parent.
+				parent.withLock(func() {
+					added := parent.addChildLocked(s.i.crdb, !opts.ParentDoesNotCollectRecording)
+					if added {
+						s.i.crdb.mu.parent = opts.Parent.i.crdb
+					}
+				})
 			}
 		}
 		s.i.crdb.enableRecording(opts.recordingType())

